### PR TITLE
Add LLVM IR view

### PR DESCRIPTION
**Pull request by RabsRincon**
This adds a llvm ir view. As a PR so my missconceptions about IR can be caught here instead of once they hit the live servers.

I'm not sure if `-fsyntax-only` is necessary, but I followed the "When in doubt, do as ~int~ast does" motto there.

I've also fixed an AST bug where it was looking at the wrong place to see if it supported ast (Request instead of compiler)

### CODE REVIEW

File: lib/base-compiler.js [hunk 1]
```diff
@@ -199,6 +199,28 @@ class BaseCompiler {
             .then(this.processAstOutput);
     }
 
+    generateIR(inputFilename, options) {
+        // These options make Clang produce an IR
+        let newOptions = _.filter(options, option => option !== '-fcolor-diagnostics')
+            .concat(this.compiler.irArg);
+
+        let execOptions = this.getDefaultExecOptions();
+        // A higher max output is needed for when the user includes headers
```
Comment (RabsRincon) [inline lib/base-compiler.js]: Needed here?

Comment (jaredwy) [reply lib/base-compiler.js]: Probably. llvm ir is pretty wordy.

File: static/panes/ir-view.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+function Ir(hub, container, state) {
+    this.container = container;
+    this.eventHub = hub.createEventHub();
+    this.domRoot = container.getElement();
+    this.domRoot.html($('#ir').html());
+    this._currentDecorations = [];
+    this.astEditor = monaco.editor.create(this.domRoot.find(".monaco-placeholder")[0], {
+        value: "",
+        scrollBeyondLastLine: false,
+        language: 'plaintext',
```
Comment (jaredwy) [inline static/panes/ir-view.js]: Reminds I have a half written llvm ir language for monaco. Should dust that off.

File: lib/llvm-ir.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            line.source = {
+                file: this.getFileName(debugInfo, line.scope),
+                line: this.getSourceLineNumber(debugInfo, line.scope)
+            };
+        });
+
+        return result;
+    }
+
+    processIrLine(line, filters) {
```
Comment (jaredwy) [inline lib/llvm-ir.js]: isn';t this just filterAsmLine now?

File: static/panes/ir-view.js [hunk 4]
```diff
@@ -85,17 +112,29 @@ Ir.prototype.initCallbacks = function () {
...
     this.eventHub.emit('requestSettings');
 
     this.container.on('resize', this.resize, this);
     this.container.on('shown', this.resize, this);
 };
 
-// TODO: de-dupe with compiler etc
+Ir.prototype.onColours = function (editorid, colours, scheme) {
+    if (editorid === this._editorid) {
+        var irColours = {};
```
Comment (jaredwy) [inline static/panes/ir-view.js]: maybe use a let here to avoid it hoising out of scope.

Comment (RabsRincon) [reply static/panes/ir-view.js]: We can't use es6 features on client side code :(

**Comment by jaredwy** [comment]
Given this a quick once over, ive put my foot down and not working this weekend :D So ill test it out tomorrow for you

**Comment by bugwelle** [comment]
@RabsRincon Thank you so much for this PR! :smiley: 
I'm "a bit" late on this, just two thoughts:

There's a glitch: If you have added the IR view and then click on the plus, the menu is only half visible (see screenshot below)

screenshot_20180826_203204

The IR syntax highlighting does not work on my machine. Will have to look into this. :)


Regards,
Andre

**Comment by mattgodbolt** [comment]
Where are we with this PR? Sounds like it needs to be updated and conflicts resolved, as a minimum. Also @bugwelle says it isn't working on his machine, perhaps we can collaborate here and get this over the finish line?

**Comment by mattgodbolt** [comment]
@bugwelle please don't worry at all! This is great. The UI glitch is not unique to your editor I think is not of your making and exists already. I _think_ @RabsRincon has some changes that might fight it across the board in his big Bootstrap 4.1 update.

For now I think the filters are fine. I'd rather get this merged and linked and then see how we feel after.

Thanks again! I think the plan is to merge @RabsRincon 's Bootstrap 4.1, then I think we should try and get this merged. Yay!

**Comment by JohanEngelen** [comment]
fyi: LDC also supports LLVM IR output. The cmdline option is `-output-ll` (instead of `-output-s`). Currently can be obtained by passing `-output-s=false -output-ll` to LDC on current CE website.

**Comment by mattgodbolt** [comment]
How does this compare to @bugwelle 's now-merged change? are we all set now?

**Pull request closed without merging**